libraries/integration/Microsoft.Bot.Builder.ApplicationInsights.Core/TelemetryBotIdInitializer.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.ApplicationInsights.Channel;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Memory;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.ApplicationInsights.Core
+{
+    /// <summary>
+    /// Initializer that sets the user ID based on Bot data.
+    /// </summary>
+    public class TelemetryBotIdInitializer : ITelemetryInitializer
+    {
+        private readonly IMemoryCache _memoryCache;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+
+        public TelemetryBotIdInitializer(IHttpContextAccessor httpContextAccessor, IMemoryCache memoryCache)
+        {
+            _memoryCache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
+            _httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
+        }
+
+        public void Initialize(ITelemetry telemetry)
```

Comment:
⚠️ If you were to keep this implementation (which I don't suggest you do)... ⚠️ 

This would now become:

```CSharp
public void Initialize(ITelemetry telemetry)
{
    var httpContext = _httpContextAccessor.HttpContext;

    if (httpContext != null && (telemetry is RequestTelemetry || telemetry is EventTelemetry))
    {
        var body = httpContext.Items[HttpContextItemsKey] as string;

        if (body != null)
        {
            // Parse the JSON stored by the companion ASP.NET middleware so we can extract values from it
            var activity = JObject.Parse(body);

            var userId = activity["from"]?["id"].Value<string>();
            var channelId = activity["channelId"].Value<string>();
            var conversationId = activity["conversation"]?["id"].Value<string>();

            var telemetryContext = telemetry.Context;

            // Set the user id on the Application Insights telemetry item.
            telemetryContext.User.Id = channelId + userId;

            // Set the session id on the Application Insights telemetry item.
            telemetryContext.Session.Id = conversationId;

            // Set the activity id
            telemetryContext.GlobalProperties.Add("activity_id", activity["id"].Value<string>());
        }
    }
}
```